share/vss/vss.go

```diff
@@ -0,0 +1,579 @@
+	// C = F + G
+	d.deals = make([]*Deal, len(d.verifiers))
+	for i := range d.verifiers {
+		fi := f.Eval(i)
+		gi := g.Eval(i)
+		d.deals[i] = &Deal{
+			SessionID:   d.sessionID,
+			SecShare:    fi,
+			RndShare:    gi,
+			Commitments: commitments,
+			T:           uint32(d.t),
+		}
+		if d.deals[i].Signature, err = sign.Schnorr(d.suite, d.long, msgDeal(d.deals[i])); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+// Deals returns a list of deal for each verifiers.
+func (d *Dealer) Deals() []*Deal {
+	var out = make([]*Deal, len(d.deals))
+	copy(out, d.deals)
+	return out
+}
+
+// ProcessComplaint analyzes the given complaint. If it's a valid complaint,
+// then it returns a Justification. This response must be broadcasted to every
+// participants. If it's an invalid complaint, it returns an error about the
+// complaints. The verifiers will also ignore an invalid Complaint.
+func (d *Dealer) ProcessComplaint(c *Complaint) (*Justification, error) {
+	if err := d.verifyComplaint(c); err != nil {
+		return nil, err
+	}
+	return &Justification{
+		// index is guaranteed to be good because of d.verifyComplaint before
+		Deal:      d.deals[int(c.Index)],
+		Complaint: c,
+	}, nil
+}
+
+// ProcessApprovals looks if the approval is legitimate and stores it. If the
+// approval is not legitimate or has already been received, it returns an error.
+// To knoew whether participants approved the sharing, call `d.EnoughApproval`
+// and if true, `d.DealCertified`.
+func (d *Dealer) ProcessApprovals(a *Approval) error {
+	return d.verifyApproval(a)
+}
+
+// SecretCommit returns the commitment of the secret being shared by this
+// dealer. This function is only to be called once the deal has enough approvals
+// and is verified otherwise it returns nil.
+func (d *Dealer) SecretCommit() abstract.Point {
+	if !d.EnoughApprovals() || !d.DealCertified() {
+		return nil
+	}
+	return d.suite.Point().Mul(nil, d.secret)
+}
+
+// Verifier receives a Deal from a Dealer, can reply by a Complaint, and can
+// collaborate with other Verifiers to reconstruct a secret.
+type Verifier struct {
+	suite abstract.Suite
+
+	long abstract.Scalar
+	pub  abstract.Point
+
+	dealer abstract.Point
+
+	index     int
+	verifiers []abstract.Point
+
+	h abstract.Point
+
+	*aggregator
+}
+
+// NewVerifier returns a Verifier out of:
+// - its longterm secret key
+// - the longterm dealer public key
+// - the list of public key of verifiers. The list MUST include the public key
+// of this Verifier also.
+// The security parameter t of the secret sharing scheme is automatically set to
+// a default safe value. If a different t value is required, it is possible to set
+// it with `verifier.SetT()`.
+func NewVerifier(suite abstract.Suite, longterm abstract.Scalar, dealerKey abstract.Point,
+	verifiers []abstract.Point) (*Verifier, error) {
+	v := &Verifier{
+		suite:     suite,
+		long:      longterm,
+		dealer:    dealerKey,
+		verifiers: verifiers,
+	}
```

Comment:
Why already creating the verifier struct at this point and not only after the check has passed in line 237?